[SofaPython] General SofaPython cleaning

matthieu-nesme opened the pull request:
EDITED: Damien

Currently the SofaPython plugin is in a very poor shapes with two serious problems. The first one is a problem for developper with a lot of duplicated or invalid or in-elegant code which impact maintainability of our code base. The second problem is the lack of consistency in the way warning and errors are reported. In sofa some component rise en exception while others prints a message and return None which is very problematic to users. 

In this PR we addressed these two issues through a major cleaning set of changes:
 - [SofaPython]: 
       * refactor the way to get get Sofa object from their PyObject equivalent (hundred of changes). 
               * add dedicated easing function (unwrap, get, get_baseobject) 
               * replace wherever possible copy-pasted code with these function. 
       * refactor the way we report error from binding (hundred of changes) (API BREAK). 
               * returning NULL with a correctly set exception type as it is the way of handling error in python. 
               * when returning NULL it is mandatory to preceed the return with an exception type & message specification. This can be done either with PyExc_SetString() or Py_BaddArgument(), 
               * when returning after a failing Py_ParseTuple is an exception, as the function sets itself a very 
descriptive and standardized error message.
               * add macros (SP_PYERR_SETSTRING_INVALIDTYPE and SP_PYERR_SETSTRING_OUTOFBOUND) to report consistant message with PyErr_SetString        
               * add SP_CLASS_METHOD_DOC() and SP_CLASS_METHOD_KW_DOC() to declare method with their docstring (if you don't know what are docstring read: http://sametmax.com/les-docstrings/)
       * add new bindings:
               * BaseState::getSize 
               * BaseMapping::applyJT
               * BaseMapping::applyDJT
               * Data::getCounter 
               * Data::isDirty
               * Node::getMass
               * Node::getForceField
               * STLExporter
               * Sofa.unload
               * Sofa.path
               * Sofa.getAvailableComponents
               * VisualModel::updateVisual
               * VisualModel::initVisual
       * add in PythonEnvironment::getPythonCallingPointAsFileInfo() a function that return a string with the 
         python file/line info to add report them in msg_* 
       * adds 'override' keyword when needed (more info => http://en.cppreference.com/w/cpp/language/override)
       * adds tests on PythonScriptController
       * adds tests on SofaModule
       * adds a exception handler in python so that the un-catched exceptions are sended as msg_error instead of being printed on the console.  
       * cosmetic changes like replacing 'using namespace a;' with the precise 'using a:TheClass', removing in-consistant line spacing. 
       * replace "extern C" with static

 - [Flexible/python] Fix the examples & the tests to take into account the changes in SofaPython
 - [Compliant/python] Fix the examples & the tests to take into account the changes in SofaPython

To simplify the submitter's life (and don't waste their time) some extras (read not really relevant) changes are also added to the PR:
  - remove a virtual in-heritance on BaseController and plugin/optitracknet. 
  - adds a << operator to Message.h to handle file:line information (now in PR: #314 ) 
  - made some changes in STLExporter to stay consistant with Sofa guidelines. 

There is still some work todo (if you have free time to offer):
Eg: 
in Compliant there is still patterns like:
```cpp
     SP_MESSAGE_ERROR( "_Compliant_getAssembledImplicitMatrix: wrong arguments" );
     return NULL;
```
The SP_MESSAGE_ERROR is probably not needed as it duplicate the one provided by the python exception. 

While in 
```cpp
   SP_MESSAGE_ERROR( "_Compliant_getAssembledImplicitMatrix: first argument is not a BaseNode" );
  PyErr_BadArgument();
  return NULL;
```
PyExc_SetString() should replace the message error & the bad argument. 

More generally there is still a lot of SP_MESSAGE_*() instead of msg_* and there is a lot of method that haven't their docstring. 
 
______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

matthieu-nesme committed:
fixing warnings

(cherry picked from commit 053c90e99f2704702784034028b51cb7a6b125db)

# Conflicts:
#	SofaKernel/modules/sofa/simulation/simulation_test/graph/DAG_test.cpp

matthieu-nesme committed:
[SofaPython] more explicit warning message about unused argument during object creation.

(cherry picked from commit fb11510180c8f07b86b4fe30cf2875ffae0a0c4d)

matthieu-nesme committed:
[SofaPython] adding a visitor to count nodes and components (to know who has the biggest of course)

(cherry picked from commit dd6997fab0c71dddcb41209b5eb2cd27ed088c36)

Commit:
[SofaPython] units API: add get functions to retrieve local units

(cherry picked from commit 04c55527d23297172b177ad8de0eb06e16bdf36a)

matthieu-nesme committed:
[SofaPython] printing the warning message for Mesh/Image not found while loading sml file only if printLog is true

(cherry picked from commit 2521fbb967782e147580d8dfea4f7b83d219cdaf)

Commit:
[SofaPython] proper error handling

(cherry picked from commit 7ad61243bf59840d6206fc896cfe69f95f87e516)

# Conflicts:
#	applications/plugins/SofaPython/Binding_GridTopology.cpp
#	applications/plugins/SofaPython/Binding_LinearSpring.cpp
#	applications/plugins/SofaPython/Binding_Link.cpp
#	applications/plugins/SofaPython/Binding_SofaModule.cpp

Commit:
[SofaPython.sml] proper comment

(cherry picked from commit d4a1cdda98b90be1d0c9c8fdbafc986adaea33e4)

matthieu-nesme committed:
[SofaPython] sml: fixing a comment

(cherry picked from commit feac032f12f8aa1a3d5b13390029a7b0f5a290f0)

Commit:
[SofaPython] better error reporting

(cherry picked from commit 6448b5a20865c8ef07c0406fbf46eb73ed33796b)

Commit:
[SofaPython] loadPlugin returns plugin path

this is useful for simple ctypes-based bindings

(cherry picked from commit de12a83a7675a0bfeb3c2dde7c2f2043fe90d41b)

Commit:
[Compliant] automatically find python tests

(cherry picked from commit 7016956c9791ae0efa5e304fbbad332e76cee3cd)

Commit:
[SofaPython] bindings for applyJT and applyDJT

(cherry picked from commit 8a36eb07d5bd1bd74872fd1b7569f8a5f58b97b2)

Commit:
[SofaPython] optional args in generateRigid

(cherry picked from commit d0defaeff014ef3785ee0138a1db8d782c560bde)

matthieu-nesme committed:
[SofaPython] propagating events from the right node

(cherry picked from commit e5301821ab54b9b285b90c4a04798148ca66f254)

matthieu-nesme committed:
[SofaPython] adding binding Data::getCounter

(cherry picked from commit 36c92cbcf266e20dfd7248c7f7bdaa7eeb320aaa)

Commit:
[SofaPython] minor

(cherry picked from commit 9b8e0f531f830526e22f9776049bbd20a9a6be9e)

matthieu-nesme committed:
[SofaPython] binding Data::isDirty

(cherry picked from commit cb4a1a0df2a855813450351723207de66c9f5ffd)

matthieu-nesme committed:
[SofaPython] adding Base::downCast to down cast to the lower type known by the factory

(cherry picked from commit 07edc94dde7ad1842dc9c66be28a9e873d08232f)

matthieu-nesme committed:
[SofaPython] minor comment fix

(cherry picked from commit c4ad1a0f7ec391df283355b1c7ecd660b6d0d4fe)

matthieu-nesme committed:
[SofaPython] sml: surfaceLinks between more that two surfaces

(cherry picked from commit a7b124c0777e95da0e2fca196df60ec99110b596)

Commit:
[SofaPython] add len() support for class Group

(cherry picked from commit 4fa7ed1ca5c66a9b4362f62761321b716ba83a66)

Commit:
[SofaPython] Quaternion : factorize some code, and add slerp function

(cherry picked from commit 61c011312b261dfd325c0831717c014d0d2ffde2)

Commit:
[Compliant] use SofaTest test list from dir

(cherry picked from commit 34e977bb68640ed3981691170a764c337ec12362)

Commit:
[SofaPython] get root node *before* calling createScene

(cherry picked from commit e14ddf0afb2eaeb8c2f71cb2ef63f22fe9e44c97)

# Conflicts:
#	applications/plugins/SofaTest/Python_test.cpp
#	applications/plugins/SofaTest/Python_test.h

Commit:
[Compliant] auto python file list in cmake

(cherry picked from commit c935bba062a75cc075b084635445b3932aac6847)

# Conflicts:
#	applications/plugins/Compliant/Compliant_test/CMakeLists.txt

Commit:
[Compliant] plugged a *huge* memleak in numpy bindings

(cherry picked from commit f815698b8ddfdbdc851674b42acc8d578a5fed41)

Commit:
[Compliant] fixed another numpy related memleak

(cherry picked from commit d6f28ad7de8989f05e6fdca54cbeb15bec44da76)

# Conflicts:
#	applications/plugins/Compliant/python/Compliant/easy.py

matthieu-nesme committed:
[SofaPython] removing an unnecessary dynamic_cast

(cherry picked from commit 3bf248ccad6c1b9ae1d8d3ab3ba24f72f56add2a)

Commit:
[SofaPython] added optional listening to all events

(cherry picked from commit 4a6460c02b2fa00f35146bb65082dbea13c3306f)

# Conflicts:
#	applications/plugins/SofaPython/Binding_PythonScriptController.cpp

Commit:
[SofaPython] fix bug in quaternion.py

(cherry picked from commit cf16385f947318cf2dec8ecfe700903817a9d65e)

Commit:
[SofaTest] put the test in the right plugin

(cherry picked from commit 3306170b006dad54e6b5f05fd45129d7f1fc1e34)

# Conflicts:
#	applications/plugins/Compliant/Compliant_test/python_test_list.cpp
#	applications/plugins/SofaTest/SofaTest_test/scenes/python_test_api.py

matthieu-nesme committed:
[SofaPython] adding __str__ implementation to Data

(cherry picked from commit 2c3d870d6a2c6b5ec18e77fcff957f6601d0675b)

matthieu-nesme committed:
[SofaPython] sml: adding Solid.keyPositions
It has been added on top of the regular 'position' not to break existing scripts.
It can be used to store useful positions, animated keyframes...

(cherry picked from commit 2dd1fc55b444c576c58414f854b725a1fe841aaf)

Commit:
[SofaPython] Sofa.path() => sofa::helper::Utils::getSofaPathPrefix()

(cherry picked from commit a75ceb7f77de3bc130af7c630c559463c7054789)

matthieu-nesme committed:
[SofaPython] fixing warning

(cherry picked from commit 050dcc86c3958d3b498c2bf201726d976d7c2f8a)

matthieu-nesme committed:
[sml] moving 'meshExporters/exportMeshes' from 'SceneArticulatedRigid' to 'BaseScene' and use it in 'SceneSkinningRigidScale'

(cherry picked from commit de7860bb26ab8fa1815690fd20999130d42b61a4)

Commit:
[SofaPython] extension methods don't need extern "C"

(cherry picked from commit 87df41864ef7bf0b78fdad90b24b005502733a6f)

matthieu-nesme committed:
[SofaPython] binding VisualModel::updateVisual()

(cherry picked from commit cc4d94ca3a12c2da3821958ddfb4cb6809c490b2)

Commit:
[SofaPython] added getSize for mstates

(cherry picked from commit 7d8ceb8fadb09b809edfe3bb22c243653051b27c)

Commit:
[SofaPython] node.getMass()

(cherry picked from commit 5a830453846451023aea097a4e18d4b25bc6275b)

Commit:
[SofaPython] bindings don't need extern "C"

(cherry picked from commit 28280767e1b9e8b2a9a7c256b75ed7e84c06e4e1)

Commit:
[SofaPython] added node.getForceField(index)

(cherry picked from commit 3d3677e51a5418e105e3afa4cb2db850bdfc1b2b)

Commit:
[SofaPython] code cleanup on Binding_Node.cpp

(cherry picked from commit 1d14256de37e8a3842a51c2f883050498479e5e4)

Commit:
[SofaPython] handle SystemExit by throwing c++ exception

(cherry picked from commit e86d2e0162edc23ff359cbd86a2fbe1346194566)

matthieu-nesme committed:
fixing warnings

(cherry picked from commit f6f062e3e452e88b0d205483b89e6a62554d2391)

# Conflicts:
#	applications/plugins/SofaTest/Python_test.cpp

matthieu-nesme committed:
[Compliant] fixing test

(cherry picked from commit 3350c5c1639daf2811f75877b8933e6549844364)

matthieu-nesme committed:
[SofaPython] binding STLExporter

(cherry picked from commit 18b777abca87b27cb56292539e8b5e353e6f22f5)

matthieu-nesme committed:
[SofaPython] binding VisualModel::initVisual

(cherry picked from commit 1f7ae6a930c26b018295ab34f9d8578e5eb3eb81)

matthieu-nesme committed:
[SofaPython] fixing units

(cherry picked from commit b1d49d9f983ee1ae2c3cef2e34c572b5932c409a)

Commit:
[SofaPython] added node.initVisual() -> VisualLoop::initStep

(cherry picked from commit 54bc7101e82a98b1d034c44b7ecec23f18db2bab)

Commit:
[SofaPython] cleanup + fixes Binding_Base.cpp

(cherry picked from commit 8bc01e119c24d62bfe16053bf8ace9b0397f9267)

Commit:
[SofaPython] extern "C" in Binding_Data

(cherry picked from commit 1e3d105e171126027453f8dc1b13b947d9b497fa)

matthieu-nesme committed:
[SofaPython] pre-importing scipy from the main thread.
(to avoid locking when importing scipy in another thread, like when loading a scene asynchronously)

(cherry picked from commit c264c3d3fdd5cbfbd49c05c795848fe9a0b9e493)

Commit:
[SofaPython] fixes + cleanup Sofa binding

(cherry picked from commit ee8347e01e08c5b62af153e85a127e98b752e326)

matthieu-nesme committed:
[SofaPython] fixing a few Py_RETURN_NONE with return NULL

matthieu-nesme commented:
I just cherry-picked Anatoscope's changes related to SofaPython (excl. PythonScriptController that comes in #283).
It seems huge, but it is simply:
- mainly cleaning-up,
- adding bindings,
- refactoring python-tests.

More cleanings are coming, and waiting for #283 and #286.

matthieu-nesme committed:
[SofaPython] replacing extern C with static

matthieu-nesme committed:
[SofaPython] fix a compilation error

matthieu-nesme committed:
[SofaPython] fixing compilation

damienmarchal commented:
@matthieu-nesme I like this PR a lot :)

matthieu-nesme committed:
[SofaPython] shared sofa getters from PyObject.

(cherry picked from commit 1774cfe4931bac3e47f3bfb7b9885038f61a4dd7)

# Conflicts:
#	applications/plugins/SofaPython/Binding_PythonScriptController.cpp

matthieu-nesme committed:
[SofaPython] binding Simulation::unload as Sofa.unload

(cherry picked from commit 50ef91cae93e07d3e25053500b47d708a2f0cfbd)

matthieu-nesme committed:
[SofaPython] more explicit memory test

(cherry picked from commit b0e2c0243243b76811f23c1722be5939b0568258)

matthieu-nesme committed:
[SofaPython] cmake: adding memory test python file

(cherry picked from commit d17a74a6d4d549afd05d9635a3bcf5183f66fd5a)

matthieu-nesme committed:
[SofaPython] cleaning controllerVariable example

(cherry picked from commit e341e2e19fec52431c54f5fef6558aa430afa8ea)

# Conflicts:
#	applications/plugins/SofaPython/examples/ControllerVariable.py

matthieu-nesme committed:
[SofaPython] minor cleaning

(cherry picked from commit b5e97defdc5cb1f9d3f1a723076fb8db853f23f0)

matthieu-nesme committed:
[OptiTrackNatNet] fixing inheritance

(cherry picked from commit 41ddb42234148d5c5009acdfe97c9837591d7fc9)

# Conflicts:
#	applications/plugins/OptiTrackNatNet/OptiTrackNatNetDevice.h

matthieu-nesme committed:
[Compliant] python: replacing extern C with static

matthieu-nesme committed:
[SofaTest] put the test in the right plugin

(cherry picked from commit 3306170b006dad54e6b5f05fd45129d7f1fc1e34)

# Conflicts:
#	applications/plugins/Compliant/Compliant_test/python_test_list.cpp
#	applications/plugins/SofaTest/SofaTest_test/scenes/python_test_api.py (reverted from commit 1a4696b3c706de99de6c2c25f19e3b3e3d14cc6d)

matthieu-nesme committed:
[SofaTest] Python_test: printing scene path on failure

bcarrez commented:
It looks huge, for sure ;-)

Commit:
[SofaPython] fixed bogus obj loader

(cherry picked from commit f25dcdf0ba723f5bf4cb7890bbf5f2bc78e3ea00)

Commit:
[SofaPython] anonymous namespace around pytypeobject patchers

otherwise link may remove duplicate symbols and disable patch :-/

(cherry picked from commit 2ffc5609d34efd8d17b349314cb358b9fa088d82)

matthieu-nesme committed:
Merge remote-tracking branch 'origin/master' into sofapython_clean

# Conflicts:
#	applications/plugins/SofaPython/Binding_Base.cpp
#	applications/plugins/SofaPython/Binding_BaseContext.cpp
#	applications/plugins/SofaPython/Binding_Data.cpp
#	applications/plugins/SofaPython/Binding_Node.cpp
#	applications/plugins/SofaPython/Binding_PythonScriptController.cpp
#	applications/plugins/SofaPython/Binding_SofaModule.cpp
#	applications/plugins/SofaPython/PythonScriptController.cpp
#	applications/plugins/SofaPython/SofaPython_test/PythonFactory_test.cpp

maxime-tournier commented on SofaKernel/framework/sofa/helper/logging/Messaging.h:
```diff
@@ -159,7 +159,7 @@
 
 #define msg_warning(...) MSGWARNING_CHOOSER(__VA_ARGS__)(__VA_ARGS__)
-#define msg_warning_when(cond, ...) if((cond)) msg_warning(__VA_ARGS__)
+#define msg_warning_when(cond, ...) if(cond) msg_warning(__VA_ARGS__)
```
put back the parenthesis, with macros you never know

maxime-tournier replied:
I mean you DO know there's gonna be issues if you don't put extra parens

maxime-tournier commented on SofaKernel/framework/sofa/helper/logging/Messaging.h:
```diff
@@ -171,7 +171,7 @@
 
 #define msg_error(...) MSGERROR_CHOOSER(__VA_ARGS__)(__VA_ARGS__)
-#define msg_error_when(cond, ...) if((cond)) msg_error(__VA_ARGS__)
+#define msg_error_when(cond, ...) if(cond) msg_error(__VA_ARGS__)
```
again

maxime-tournier commented on SofaKernel/framework/sofa/helper/logging/Messaging.h:
```diff
@@ -183,7 +183,7 @@
 
 #define msg_fatal(...) MSGFATAL_CHOOSER(__VA_ARGS__)(__VA_ARGS__)
-#define msg_fatal_when(cond, ...) if((cond)) msg_fatal(__VA_ARGS__)
+#define msg_fatal_when(cond, ...) if(cond) msg_fatal(__VA_ARGS__)
```
again

maxime-tournier commented on SofaKernel/framework/sofa/helper/logging/Messaging.h:
```diff
@@ -195,7 +195,7 @@
 
 #define msg_deprecated(...) MSGDEPRECATED_CHOOSER(__VA_ARGS__)(__VA_ARGS__)
-#define msg_deprecated_when(cond, ...) if((cond)) msg_deprecated(__VA_ARGS__)
+#define msg_deprecated_when(cond, ...) if(cond) msg_deprecated(__VA_ARGS__)
```
here

maxime-tournier commented on applications/plugins/Compliant/python/Binding_AssembledSystem.cpp:
```diff
@@ -5,15 +5,16 @@
 
 
+
 PyObject* getMatrice( const AssembledSystem::rmat& A )
```
"matrice" ?

maxime-tournier commented on applications/plugins/Compliant/python/Binding_AssembledSystem.cpp:
```diff
@@ -32,9 +33,9 @@ PyObject* getMatrice( const AssembledSystem::rmat& A )
+    AssembledSystem* sys = get<AssembledSystem>( self );
     if (!sys)
     {
         PyErr_BadArgument();
```
`return NULL` on next line

maxime-tournier commented on applications/plugins/Compliant/python/Binding_AssembledSystem.cpp:
```diff
@@ -44,9 +45,9 @@ extern "C" PyObject* AssembledSystem_getH(PyObject * self, PyObject * /*args*/)
+    AssembledSystem* sys = get<AssembledSystem>( self );
     if (!sys)
     {
         PyErr_BadArgument();
```
again

maxime-tournier commented on applications/plugins/Compliant/python/Binding_AssembledSystem.cpp:
```diff
@@ -56,9 +57,9 @@ extern "C" PyObject* AssembledSystem_getP(PyObject * self, PyObject * /*args*/)
+    AssembledSystem* sys = get<AssembledSystem>( self );
     if (!sys)
     {
         PyErr_BadArgument();
```
again

maxime-tournier commented on applications/plugins/Compliant/python/Binding_AssembledSystem.cpp:
```diff
@@ -68,9 +69,9 @@ extern "C" PyObject* AssembledSystem_getJ(PyObject * self, PyObject * /*args*/)
+    AssembledSystem* sys = get<AssembledSystem>( self );
     if (!sys)
     {
         PyErr_BadArgument();
```
again

maxime-tournier commented on applications/plugins/Compliant/python/Binding_AssembledSystem.cpp:
```diff
@@ -94,7 +95,7 @@ SP_CLASS_METHODS_END
+    AssembledSystem* obj = get<AssembledSystem>( self );
     if (!obj)
     {
         PyErr_BadArgument();
```
again

maxime-tournier commented on applications/plugins/Compliant/python/Binding_AssembledSystem.cpp:
```diff
@@ -104,7 +105,7 @@ SP_CLASS_ATTR_GET(AssembledSystem,m)(PyObject *self, void*)
+    AssembledSystem* obj = get<AssembledSystem>( self );
     if (!obj)
     {
         PyErr_BadArgument();
```
IIRC this should be `-1`

maxime-tournier replied:
(the return value)

maxime-tournier commented on applications/plugins/Compliant/python/Binding_AssembledSystem.cpp:
```diff
@@ -116,7 +117,7 @@ SP_CLASS_ATTR_SET(AssembledSystem,m)(PyObject *self, PyObject * args, void*)
+    AssembledSystem* obj = get<AssembledSystem>( self );
     if (!obj)
     {
         PyErr_BadArgument();
```
`return NULL;` next line

maxime-tournier commented on applications/plugins/Compliant/python/Binding_AssembledSystem.cpp:
```diff
@@ -126,7 +127,7 @@ SP_CLASS_ATTR_GET(AssembledSystem,n)(PyObject *self, void*)
+    AssembledSystem* obj = get<AssembledSystem>( self );
     if (!obj)
     {
         PyErr_BadArgument();
```
IIRC `return -1;` next line

maxime-tournier commented on applications/plugins/Compliant/python/Binding_AssembledSystem.cpp:
```diff
@@ -138,7 +139,7 @@ SP_CLASS_ATTR_SET(AssembledSystem,n)(PyObject *self, PyObject * args, void*)
+    AssembledSystem* obj = get<AssembledSystem>( self );
     if (!obj)
     {
         PyErr_BadArgument();
```
again

maxime-tournier commented on applications/plugins/Compliant/python/Binding_AssembledSystem.cpp:
```diff
@@ -148,7 +149,7 @@ SP_CLASS_ATTR_GET(AssembledSystem,dt)(PyObject *self, void*)
+    AssembledSystem* obj = get<AssembledSystem>( self );
     if (!obj)
     {
         PyErr_BadArgument();
```
again

maxime-tournier commented on applications/plugins/Compliant/python/Compliant/easy.py:
```diff
@@ -25,84 +32,110 @@ class Opaque(Structure):
+dll.eigen_sizeof.restype = c_size_t
+dll.eigen_sizeof.argtypes = ()
 
+def as_buffer(ptr, *size):
```
comment:
```'''cast a ctypes pointer as a correctly typed (multidimensional) array of given `size`

this is needed to avoid a nasty memleak with numpy.ctypeslib.as_array
'''```

maxime-tournier commented on applications/plugins/Compliant/python/_Compliant.cpp:
```diff
@@ -39,7 +40,7 @@ extern "C" PyObject * _Compliant_getAssembledImplicitMatrix(PyObject * /*self*/,
         Py_RETURN_NONE;
```
don't `PyErr_BadArgument`, just `return NULL` directly

maxime-tournier replied:
`PyArg_ParseTuple` already sets exception

maxime-tournier commented on applications/plugins/Compliant/python/_Compliant.cpp:
```diff
@@ -95,7 +96,7 @@ extern "C" PyObject * _Compliant_getImplicitAssembledSystem(PyObject * /*self*/,
         Py_RETURN_NONE;
```
again

maxime-tournier commented on applications/plugins/Compliant/python/_Compliant.cpp:
```diff
@@ -135,15 +136,15 @@ extern "C" PyObject * _Compliant_getLambda(PyObject * /*self*/, PyObject * args)
         Py_RETURN_NONE;
```
and again

maxime-tournier commented on applications/plugins/Compliant/python/_Compliant.cpp:
```diff
@@ -135,15 +136,15 @@ extern "C" PyObject * _Compliant_getLambda(PyObject * /*self*/, PyObject * args)
     {
         SP_MESSAGE_ERROR( "_Compliant_getConstraintForce: wrong arguments - not a CompliantImplicitSolver" );
         PyErr_BadArgument();
         Py_RETURN_NONE;
```
and again

maxime-tournier replied:
i mean `PyErr_BadArguement` is ok here, but you need to `return NULL`;

maxime-tournier commented on applications/plugins/Compliant/python/_Compliant.cpp:
```diff
@@ -183,15 +184,15 @@ extern "C" PyObject * _Compliant_propagateLambdas(PyObject * /*self*/, PyObject
         Py_RETURN_NONE;
```
and again

maxime-tournier commented on applications/plugins/SofaPython/Binding_BaseContext.cpp:
```diff
@@ -39,7 +40,7 @@ using namespace sofa::defaulttype;
 
 extern "C" PyObject * BaseContext_setGravity(PyObject *self, PyObject * args)
```
`static`

maxime-tournier commented on applications/plugins/SofaPython/Binding_BaseContext.cpp:
```diff
@@ -39,7 +40,7 @@ using namespace sofa::defaulttype;
+    BaseContext* obj = get_basecontext( self );
     PyPtr<Vector3>* pyVec;
     if (!PyArg_ParseTuple(args, "O",&pyVec))
         Py_RETURN_NONE;
```
return NULL;

maxime-tournier commented on applications/plugins/SofaPython/Binding_BaseMapping.cpp:
```diff
@@ -71,7 +72,7 @@ extern "C" PyObject * BaseMapping_setFrom(PyObject * self, PyObject * args)
         Py_RETURN_NONE;
```
return NULL;

maxime-tournier commented on applications/plugins/SofaPython/Binding_BaseMapping.cpp:
```diff
@@ -84,15 +85,15 @@ extern "C" PyObject * BaseMapping_setFrom(PyObject * self, PyObject * args)
 
     PyObject* pyTo;
     if (!PyArg_ParseTuple(args, "O",&pyTo))
         Py_RETURN_NONE;
```
return NULL;

matthieu-nesme committed:
[SofaPython] fixing compilation and warnings

maxime-tournier commented on applications/plugins/SofaPython/Binding_BaseState.cpp:
```diff
@@ -22,28 +22,46 @@
-    if (!PyArg_ParseTuple(args, "i",&newSize))
-        Py_RETURN_NONE;
+    if (!PyArg_ParseTuple(args, "i", &newSize)) {
+        PyErr_BadArgument();
```
No `PyErr_BadArgument`

maxime-tournier commented on applications/plugins/SofaPython/Binding_BaseState.cpp:
```diff
@@ -22,28 +22,46 @@
+    BaseState* obj = get_basestate( self );
+    
+    if (PyTuple_Size(args)) {
+        PyErr_BadArgument();
```
again

maxime-tournier commented on applications/plugins/SofaPython/Binding_Data.cpp:
```diff
@@ -561,9 +563,9 @@ extern "C" PyObject * Data_getSize(PyObject *self, PyObject * /*args*/)
+    BaseData* data = get_basedata( self );
     int size;
     if (!PyArg_ParseTuple(args, "i",&size))
     {
```
again next line

maxime-tournier commented on applications/plugins/SofaPython/Binding_Data.cpp:
```diff
@@ -576,28 +578,28 @@ extern "C" PyObject * Data_setSize(PyObject *self, PyObject * args)
+    BaseData* data = get_basedata( self );
 
     PyObject *value;
     if (!PyArg_ParseTuple(args, "O",&value))
```
again next line

maxime-tournier commented on applications/plugins/SofaPython/Binding_Data.cpp:
```diff
@@ -620,9 +622,9 @@ extern "C" PyObject * Data_read(PyObject *self, PyObject * args)
+    BaseData* data = get_basedata( self );
 
     PyObject *value;
     if (!PyArg_ParseTuple(args, "O",&value))
```
again

maxime-tournier commented on applications/plugins/SofaPython/Binding_Data.cpp:
```diff
@@ -747,4 +773,11 @@ SP_CLASS_ATTR(Data,name)
 SP_CLASS_ATTRS_END
 
-SP_CLASS_TYPE_BASE_PTR_ATTR(Data,BaseData)
+
```
need anonymous namespace { } around patcher (see commit on anatoscope master)

maxime-tournier replied:
otherwise if there are two classes named like this in this namespace (which may happen) the linker will remove one and the constructor won't get called

maxime-tournier commented on applications/plugins/SofaPython/Binding_GridTopology.cpp:
```diff
@@ -22,13 +22,22 @@
+    GridTopology* obj = get_GridTopology( self );
     int nx,ny,nz;
     if (!PyArg_ParseTuple(args, "iii",&nx,&ny,&nz))
     {
```
no `badargument` next line

maxime-tournier commented on applications/plugins/SofaPython/Binding_GridTopology.cpp:
```diff
@@ -58,15 +68,15 @@ extern "C" PyObject * GridTopology_setNx(PyObject *self, PyObject * args)
+    GridTopology* obj = get_GridTopology( self );
     int nb;
     if (!PyArg_ParseTuple(args, "i",&nb))
     {
```
again

maxime-tournier commented on applications/plugins/SofaPython/Binding_GridTopology.cpp:
```diff
@@ -77,15 +87,15 @@ extern "C" PyObject * GridTopology_setNy(PyObject *self, PyObject * args)
+    GridTopology* obj = get_GridTopology( self );
     int nb;
     if (!PyArg_ParseTuple(args, "i",&nb))
     {
```
again

maxime-tournier commented on applications/plugins/SofaPython/Binding_LinearSpring.cpp:
```diff
@@ -21,14 +21,23 @@
+    LinearSpringR* obj = get_LinearSpringR( self );
     if (!obj)
     {
         PyErr_BadArgument();
```
this is a logic error (should never happen), maybe an assert is enough

maxime-tournier commented on applications/plugins/SofaPython/Binding_PythonScriptController.cpp:
```diff
@@ -50,7 +50,7 @@ extern "C" PyObject * PythonScriptController_onIdle(PyObject * /*self*/, PyObjec
+static PyObject * PythonScriptController_onLoaded(PyObject * /*self*/, PyObject * args)
 {
     PyObject *pyNode;
     if (!PyArg_ParseTuple(args, "O",&pyNode))
```
bad arg next few lines

maxime-tournier commented on applications/plugins/SofaPython/Binding_TopologyChange.cpp:
```diff
@@ -48,7 +56,7 @@ SP_CLASS_ATTR_SET(PointAncestorElem, type)(PyObject *self, PyObject * args, void
+    PointAncestorElem* obj = get_PointAncestorElem( self );
     if (!obj)
     {
         PyErr_BadArgument();
```
return -1; ?

maxime-tournier commented on applications/plugins/SofaPython/Binding_TopologyChange.cpp:
```diff
@@ -16,14 +24,14 @@ PyObject * PointAncestorElem_PyNew(PyTypeObject * /*type*/, PyObject * /*args*/,
+    PointAncestorElem* obj = get_PointAncestorElem( self );
     if (!obj)
     {
         PyErr_BadArgument();
```
return -1; ?

maxime-tournier commented on applications/plugins/SofaPython/Binding_TopologyChange.cpp:
```diff
@@ -60,7 +68,7 @@ SP_CLASS_ATTR_GET(PointAncestorElem,index)(PyObject *self, void*)
+    PointAncestorElem* obj = get_PointAncestorElem( self );
     if (!obj)
     {
         PyErr_BadArgument();
```
return -1; ?

maxime-tournier commented on applications/plugins/SofaPython/Binding_TopologyChange.cpp:
```diff
@@ -72,7 +80,7 @@ SP_CLASS_ATTR_SET(PointAncestorElem, index)(PyObject *self, PyObject * args, voi
+    PointAncestorElem* obj = get_PointAncestorElem( self );
     if (!obj)
     {
         PyErr_BadArgument();
```
return -1; ?

maxime-tournier commented on applications/plugins/SofaPython/Binding_TopologyChange.cpp:
```diff
@@ -91,7 +99,7 @@ SP_CLASS_ATTR_GET(PointAncestorElem, localCoords)( PyObject *self, void* )
+    PointAncestorElem* obj = get_PointAncestorElem( self );
     if (!obj)
     {
         PyErr_BadArgument();
```
again

maxime-tournier commented on applications/plugins/SofaPython/Binding_VisualModel.cpp:
```diff
@@ -24,14 +24,24 @@
+    VisualModelImpl* obj = get_VisualModelImpl( self );
     double r,g,b,a;
     if (!PyArg_ParseTuple(args, "dddd",&r,&g,&b,&a))
     {
```
bad arg next lines

maxime-tournier replied:
also it should clear the error flag on the ok path

maxime-tournier commented on applications/plugins/SofaPython/PythonToSofa.inl:
```diff
@@ -0,0 +1,153 @@
+
```
make them all `static` just in case

maxime-tournier reviewed (commented):
- don't `PyErr_BadArgument` after unsuccessful `PyArg_ParseTuple`
- `return NULL;` on errors, except in attribute setters where it should be `-1` on error (IIRC)

matthieu-nesme committed:
[SofaPython] cleaning controllerVariable example

Commit:
[SofaPython] got parenthesis back in macros

Commit:
[Compliant] python error handling

Commit:
[SofaPython] extern "C" -> static

Commit:
[SofaPyton] error handling

Commit:
[Compliant] python error handling

Commit:
[Compliant] comments

Commit:
[SofaPython] empty args parse

Commit:
[SofaPython] error handling

Commit:
[SofaPython] badargument fix

Commit:
[SofaPython] badarg

Commit:
[SofaPython] macro-injected extern "C" (wtf)

Commit:
[SofaPython] setattr error handling

Commit:
[SofaPython] badarg

Commit:
[SofaPython] clear error flag

Commit:
[SofaPython] static inline + include guards

Commit:
[SofaPython] fixed compilation

damienmarchal commented on applications/plugins/SofaPython/PythonScriptController.h:
```diff
@@ -133,6 +135,7 @@ class SOFA_SOFAPYTHON_API PythonScriptController : public ScriptController
     PyObject *m_Func_onMouseWheel;
     PyObject *m_Func_onGUIEvent;
     PyObject *m_Func_onScriptEvent;
+    PyObject* m_Func_onEvent;
```
What is this one for ? 
It adds a new onEvent, a Data<bool> and a hook in the onScriptEvent function. 
Shouldn't it be in a separated PR ?

maxime-tournier replied:
You are right, i'll remove it from this PR

maxime-tournier replied:
for the record: this is for handling arbitrary events directly from python controllers: events are passed as an opaque pointer and python code can access it using ctypes.

I just removed it from this PR.

Commit:
[SofaPython] removed PythonScriptController c event handling

bcarrez commented on modules/SofaExporter/STLExporter.h:
```diff
@@ -56,15 +56,16 @@ class SOFA_EXPORTER_API STLExporter : public core::objectmodel::BaseObject
+    void writeSTLBinary();
+
     sofa::core::objectmodel::DataFileName stlFilename;
-    Data<bool> m_fileFormat;      //0 for Ascii Formats, 1 for Binary File Format
```
Though I agree to name this var correctly, I don't see its relevance to this particular PR ? But I guess it's ok to let this change in this PR, doing a separate PR just for it would be a useless overhead...

matthieu-nesme replied:
yeah that is the problem of git, you can only cherry-pick a full commit...

damienmarchal committed:
[Python(Main)ScriptController] Add proper initialization with nullptr & add override

In this commit I:
  - added 'override' all around the virtual function in the two class.
  - added {nullptr} initializer. No doing so is the source of crash when
    there the loadFile cannot succeed
    eg of scene: <PythonScriptController noclass='AValidClass'/>
  - added a test to validate that the behavior is working.

damienmarchal committed:
Merge branch 'sofapython_clean' of github.com:sofa-framework/sofa into sofapython_clean

# Conflicts:
#	applications/plugins/SofaPython/PythonScriptController.h

damienmarchal committed:
[SofaPython] remove commented code, extra lines, replace Py_Return_None, adds TODO(PR:304)

Please use TODO(PR:304) so that we can quickly fin question or stuff to do in this PR.

damienmarchal committed:
[SofaPython/Binding] Cosmetic.

damienmarchal committed:
[SofaPython] Cosmetic

damienmarchal committed:
[SofaPyhon/BaseMapping] Replace BadArument with adequate a TypeError & SetString.

Maxime is this what should be done ?

damienmarchal committed:
[SofaPython/Binding_BaseMechanicalState.cpp] Remove BadArgument when using Py_ParseTuple

damienmarchal committed:
[SofaPython] Cosmetic, remove or replace using namespace with a more specific 'using'

Commit:
[SofaPython] eased unwrapping of pysptr/pyptr

damienmarchal committed:
[SofaPython] Cosmetic & remove PyErr_BadArrgument when Py_ParseTuple.

Commit:
[SofaPython] add wrapped type to unwrap traits + make get_arg use it

damienmarchal committed:
Merge branch 'sofapython_clean' of github.com:sofa-framework/sofa into sofapython_clean

damienmarchal committed:
[SofaPython] Still cleaning.

matthieu-nesme committed:
[SofaPython] cleaning argument unwrapping (using a dynamic_cast + nullity test)

matthieu-nesme committed:
[SofaPython] removing PyErr_BadArgument and unnecessary nullity checks on self.

damienmarchal committed:
[PythonScriptController] Replace SP_MESSAGE_* with msg_*

PythonScriptController is a sofa component so it should use
the msg_info(), msg_error() API that automatically add the component
type & name as message source instead of SofaPython.

damienmarchal committed:
[Sofapython] Add default initialization for attributes.

damienmarchal committed:
[Sofapython] Fix a typo in module description.

damienmarchal committed:
[SofaPython] Replace 'using namespace' with 'using' when appropriate.

damienmarchal committed:
[SofaPython] Cosmetic cleaning.

damienmarchal committed:
[PythonEnvironment] Remove commented code.

damienmarchal committed:
[PythonFactory] Move code from ".h" to ".cpp"

damienmarchal committed:
SofaPython] Cosmetic (make spaces between function consistant)

damienmarchal committed:
[ScriptController] Add override to virtual functions.

damienmarchal committed:
[SofaPython] Cosmetic + replace 'using namespace' by 'using'

damienmarchal committed:
[PythonMainScriptController] It is a sofa compnent...so its message shouldn't be in "SofaPython".

damienmarchal committed:
Revert "[SofaPython] removing PyErr_BadArgument and unnecessary nullity checks on self."

This reverts commit cb77ab4930567f34781864fd4d9d552a052a6edb.

damienmarchal committed:
Revert "[SofaPython] cleaning argument unwrapping (using a dynamic_cast + nullity test)"

This reverts commit 228be20b1fa6e06a4c63fd6c0d4fd8d4126396e3.

damienmarchal committed:
[SofaPython] Add tests on PythonScriptController

damienmarchal committed:
[SofaKernel] Adds a way to modify the FileInfo.

damienmarchal committed:
[SofaPython] Cosmetic changes.

damienmarchal committed:
[SofaPython/PythonMacro] Add a new macro to centralize the messges.

damienmarchal committed:
[SofaPython] Change way we emit message.

msg_* allows to bind the message to a sofa component.
So that they are displayed in the GUI and can be queried.

So I replaced when possible SP_MESSAGE_* with the right msg_* API to
restore this functionnality.

damienmarchal committed:
[SofaPython] Add more test on how we handle exception&message

damienmarchal committed:
[SofaPython] Improve the tests.

damienmarchal committed:
[SofaPython] Add a function to get the python calling point and add it into SofaPython messages.

damienmarchal committed:
[SofaPython] Replace SP_MESSAGE_WARNING... with msg_warning

damienmarchal committed:
[SofaPython] Add more test cases.

damienmarchal committed:
[SofaPython] More tests.

damienmarchal committed:
[SofaPython] Replace msg_warning("") with exception.

If the message is associated with a python expection (so preceeding a return NULL);
It is better to use the PyErr_SetString() to set the message. If not catched by python
code it will be then streamed as a msg_error message with the stacktrace.

damienmarchal committed:
[SofaPython] One more test.

TODO: I will factor that soon.

damienmarchal committed:
[SofaPython] msg_warning are not needed when returning a python exception.

Commit:
[SofaPython] fixed wrap/unwrap

Commit:
[SofaPython] unwrap all the things !11

Commit:
[SofaPython] harmonized all data bindings to PyPtr<BaseData>

Commit:
[SofaPython] added unwrap_self that static_casts

maxime-tournier commented:
A little update:
- added `wrap/unwrap` mechanism: there are wrapping traits in `PythonToSofa.inl` specifying wrapped type (`PySPtr` or `PyPtr`) for a given type (defaults to `PyPtr`)
- wrapping traits specializations for `Base`, `BaseData` and `BaseLink`-derived classes. `Base`-derived classes are wrapped as `PySPtr<Base>`, `BaseData`-derived as `PyPtr<BaseData>`, etc. this should cover most cases.
- `unwrap<T>(py_obj)` first recovers the wrapped type and the object pointer, then `dynamic_casts<T*>` 
- `unwrap_self<T>(py_obj)` does the same but with `static_cast` (this is for `self` arguments, python enforces that `self` is an instance of this type).

Extension code should look like this now:

```c++
Node* node = sofa::py::unwrap_self<Node>(self);
BaseObject* obj = sofa::py::unwrap<BaseObject>(py_obj); if(!obj) { ... }
```

damienmarchal committed:
[SofaPython] Fix test failure on centos.

It is possible that the clang version < 3.7 & gcc < 4.8 does not fully
support c++x11 regex so I implemented the string replacement "by hand".
Otherwise this code is failing on the CI.

damienmarchal committed:
[PythonScriptController_test] Refactor a bit the test.

damienmarchal committed:
[SofaKernel] Add proper handling of message FileInfo

damienmarchal committed:
[SofaPython] WIP Improve the way we returns FileInfo in SofaPython messages

damienmarchal committed:
[SofaPython] Cosmetic

damienmarchal committed:
[SofaPython] Add support for "docstrings" in our binding.

damienmarchal committed:
[SofaPython] Add docstring for BaseContext's method.

Instead of keeping the latex document up to date with the binding.
Please put energy to document python code with docstring. This is the
pythonic way to do.

damienmarchal committed:
[SofaPython] Add extra arguments in Sofa.msg_error to specify the file:lineno location

If not provided. It is deduced from the PythonEnvironment.

damienmarchal committed:
[SofaPython] Compilation problem in BaseContext due to the docstring.

damienmarchal committed:
[SofaPython] Fix parsing problem in parse_emitter_message_then.

damienmarchal committed:
[SofaPython] Fix segfaul in getPythonCallingPointString()

damienmarchal committed:
[SofaPython] adds a new test class for SofaModule.

damienmarchal committed:
[SofaPython] Fix a problem in getPythonCallingPoint

damienmarchal committed:
[SofaPython_test] The sys.except_hook is not working nicely with other sys.hook.

In tests, the hook is installed multiple time and fallback to the default hook
screwing the one installed by sofa. This commit fix the issue by chainging the hook
with the previous one.

damienmarchal committed:
[SofaKernel] Add missing SOFA_HELPER_API

damienmarchal committed:
[SofaKernel] Fix mis-placed SOFA_HELPER_API

damienmarchal committed:
Merge branch 'master' into sofapython_clean

damienmarchal commented:
@matthieu-nesme and @maxime-tournier  do you agree if we cut this PR in four on a per files basis instead of per commit ?
It could be:
- Sofa base
- SofaPython
- Flexible
- Compliant

bcarrez commented:
[ci-build]

damienmarchal commented:
@hugtalbot To me this PR is a prefect example of where we could "relax/ease" the merging process. 

From the submitter point of view we are "giving" to the community more than 50h of work and then we are requested to do more to split the PR. At some point we need to be a bit more pragmatic, having a "industrial grade" reviewing process is great, but do we have the budget for that ?

In this PR, mathieu, maxime and I contributed...if it breaks something in master, which is a development branch anyway, ...we will be there to fix it a-posteriori. 

But you are right, this PR should have a better description and this will be done when it will be taggued to review :)

maxime-tournier commented:
@damienmarchal ok for splitting, this is getting too big anyways :-)

damienmarchal commented:
Ok, I will do that this evening.

matthieu-nesme commented:
I am not sure it is doable as things must depend on each other.

damienmarchal commented:
[ci-build]

damienmarchal committed:
[SofaPython] Replace the use of PyObject_Str by PyObject_Repr when passing an object to Sofa

damienmarchal commented:
@hugtalbot, @guparan  
I really would like to take the easy path of merging the whole PR. 

This PR can break compatibility...but to me this important to allow breaking cleaning (and to have that done in a single step is far easier (in term of extra management work) than doing that in multiple small PR.

damienmarchal commented:
@hugtalbot  No it is not..sorry for that.. I will fix the conflict and the failing test.

damienmarchal committed:
Merge branch 'master' into sofapython_clean

damienmarchal committed:
[PythonFactory_test] Update the test to the replacement of str by repr

damienmarchal commented:
@hugtalbot I fixed the test, is was not updated to handle the change between repr and str as well as the commit.

damienmarchal commented:
My bad (again). I forgot to push to upstream and not origin.
